### PR TITLE
Add RemoteCode

### PR DESCRIPTION
**Pull request by chiting**
Add remote code model to fetch the remote code when order is used with `instant-url` as delivery format

### CODE REVIEW

File: lib/wegift/models/remote_code.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+    def initialize(url:)
+      @url = url
+    end
+
+    def self.get(url)
+      uri = URI(url)
+      conn = Faraday.new(url: "#{uri.scheme}://#{uri.host}")
+
+      response =  conn.get("#{uri.path}?format=json") do |req|
```
Comment (kendrikat) [inline lib/wegift/models/remote_code.rb:22]: Is this some weird requirement on their side? So the Content-Type header is ignored? Or do they actually need both?

File: lib/wegift/models/remote_code.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    attr_reader :amount, :barcode_format, :barcode_string, :code, :expiry_date,
+      :pin, :type
+
+    def initialize(url:)
+      @url = url
+    end
+
+    def self.get(url)
+      uri = URI(url)
+      conn = Faraday.new(url: "#{uri.scheme}://#{uri.host}")
```
Comment (kendrikat) [inline lib/wegift/models/remote_code.rb:20]: From the Tests I could see it's on wegift.io - why not use the Wegift::Client here? You can pass @api_path as empty string the URL as `public/gifts/instant`

File: lib/wegift/models/remote_code.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      super
+
+      if is_successful?
+        @amount = @payload['e_code']['amount']
+        @barcode_format = @payload['e_code']['barcode_format']
+        @barcode_string = @payload['e_code']['barcode_string']
+        @code = @payload['e_code']['code']
+        @expiry_date = @payload['e_code']['expiry_date']
+        @pin = @payload['e_code']['pin']
+        @type = @payload['e_code']['type']
```
Comment (kendrikat) [inline lib/wegift/models/remote_code.rb:42]: Why not just `Wegift::RemoteCode.new(@payload['e_code'])`? And why does this response has everything embedded in a `e_code`?

**Review by kendrikat — CHANGES_REQUESTED** [review]
I don't really see a reason to not use the functions this client provides. Any reason for that?

File: lib/wegift/models/remote_code.rb (hunk 1 above)
Comment (chiting) [reply lib/wegift/models/remote_code.rb:22]: I think they are trying to shoe-horn an existing endpoint (public remote voucher) and added the json version of it later.

**Comment by chiting** [comment]
> I don't really see a reason to not use the functions this client provides. Any reason for that?

The endpoint called is a public one and I wouldn't send credentials there.

File: lib/wegift/models/remote_code.rb (hunk 2 above)
Comment (chiting) [reply lib/wegift/models/remote_code.rb:20]: we could but as the endpoint is public, there's no need to send credentials there and I decided not to

File: lib/wegift/models/remote_code.rb (hunk 3 above)
Comment (chiting) [reply lib/wegift/models/remote_code.rb:42]: this response is basically a voucher redeemed. you can look at the response from their api and it seems they just tried to shoe-horn it to get rid of the old order workflow

File: lib/wegift/models/remote_code.rb (hunk 2 above)
Comment (chiting) [reply lib/wegift/models/remote_code.rb:20]: also, we can't guarantee the URL returned will always be pointing to the same as we've set for the `Wegift::Client`

File: lib/wegift/models/remote_code.rb (hunk 1 above)
Comment (kendrikat) [reply lib/wegift/models/remote_code.rb:22]: So this is required or not?

File: lib/wegift/models/remote_code.rb (hunk 2 above)
Comment (kendrikat) [reply lib/wegift/models/remote_code.rb:20]: we could sending the auth optional - I think that's less effort than rewriting the request - especially when we implement new non authed endpoints. 
And about the URL, is that somewhere documented?

File: lib/wegift/models/remote_code.rb (hunk 3 above)
Comment (kendrikat) [reply lib/wegift/models/remote_code.rb:42]: Still we can simplidy and do:
```
    def parse(response)
      super

      Wegift::RemoteCode.new(@payload['e_code'])
    end
```

**Comment by kendrikat** [comment]
> > I don't really see a reason to not use the functions this client provides. Any reason for that?
> 
> The endpoint called is a public one and I wouldn't send credentials there.

Making it optional is less effort.

File: lib/wegift/models/remote_code.rb (hunk 1 above)
Comment (chiting) [reply lib/wegift/models/remote_code.rb:22]: no required

File: lib/wegift/models/remote_code.rb (hunk 2 above)
Comment (chiting) [reply lib/wegift/models/remote_code.rb:20]: the URL is not documented, we only use what they send back to us when passing `url-instant` for the delivery format

File: lib/wegift/models/remote_code.rb (hunk 3 above)
Comment (kendrikat) [reply lib/wegift/models/remote_code.rb:42]: Don't forget to return it on line 29

File: lib/wegift/models/remote_code.rb (hunk 2 above)
Comment (chiting) [reply lib/wegift/models/remote_code.rb:20]: the only pieces of documentation we have is attached to the story card (8701)

File: lib/wegift/models/remote_code.rb (hunk 3 above)
Comment (chiting) [reply lib/wegift/models/remote_code.rb:42]: I don't understand why we need to instantiate a whole new class to set some attributes. Moreover, the new instance of `Wegift::RemoteCode` would have its `is_successful?` method returning false and we are using it to handle failure up the call chain.

File: lib/wegift/models/remote_code.rb (hunk 1 above)
Comment (chiting) [reply lib/wegift/models/remote_code.rb:22]: sorry, didn't understand your first comment. `Content-Type` is not required but for the sake of validity, we're adding it to the request.

Comment (kendrikat) [reply lib/wegift/models/remote_code.rb:18]: I guess the header is not needed than?

**Pull request merged**